### PR TITLE
Add docs for events and actions fetched from network API

### PR DESCRIPTION
**Pull request by MartinMinkov**

### CODE REVIEW

File: docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx [hunk 1]
```diff
@@ -99,6 +99,42 @@ this.reducer.getActions({ fromActionsHash?: Field, endActionsHash?: Field }): A[
 
 Right now, `getActions` is available for [testing with `LocalBlockchain`](/zkapps/how-to-test-a-zkapp).
 
+### Fetching Actions from an Archive Node
+
+Since are emitted action data is stored in transaction data, actions will only be persisted by an Archive Node. To fetch action data from an Archive Node, you can use the `archive` property on the `Mina.Network` function. For example:
```
Comment (jasongitmail) [inline docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx:104]: Grammar on first sentence.

Is the `archive` parameter optional if their app doesn't need to consume this, or will it error?

> If your smart contract uses actions, you'll need to provide an `archive` property on your object passed to `Mina.Network()`, containing an Archive Node API GraphQL URL.

File: docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx [hunk 2]
```diff
@@ -99,6 +99,42 @@ this.reducer.getActions({ fromActionsHash?: Field, endActionsHash?: Field }): A[
...
+
+```ts
+let Network = Mina.Network({
```
Comment (jasongitmail) [inline docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx:107]: const

File: docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx [hunk 3]
```diff
@@ -99,6 +99,42 @@ this.reducer.getActions({ fromActionsHash?: Field, endActionsHash?: Field }): A[
...
+  mina: '', // Use https://proxy.berkeley.minaexplorer.com/graphql or https://api.minascan.io/node/berkeley/v1/graphql
+  archive: '', // Use https://api.minascan.io/archive/berkeley/v1/graphql/ or https://archive.berkeley.minaexplorer.com/
+});
+Mina.setActiveInstance(Network);
+```
+
+By setting the `archive` property to one of the listed endpoints, SnarkyJS will use one of the selected endpoints to fetch action data. For example, let us assume that a zkApp has already emitted actions and has broadcasted the transaction. To fetch actions for a zkApp, you can use the `getActions` function:
```
Comment (jasongitmail) [inline docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx:114]: Would add parenthesis `getActions()`. But should it be `fetchActions()` instead of `getActions()`?

Will let Barrie review for wording here.

File: docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx [hunk 4]
```diff
@@ -99,6 +99,42 @@ this.reducer.getActions({ fromActionsHash?: Field, endActionsHash?: Field }): A[
...
+
+```ts
+class MyContract extends SmartContract {
+  ...
+  @method getActionsExample() {
+    // Get all actions for this zkApp
+    let pendingActions = this.reducer.getActions({
```
Comment (jasongitmail) [inline docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx:121]: cons

File: docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx [hunk 5]
```diff
@@ -99,6 +99,42 @@ this.reducer.getActions({ fromActionsHash?: Field, endActionsHash?: Field }): A[
...
+      (state: Field, action: Field) => {
+        ...
+      },
+      { state: counter, actionsHash }
+    );
+  }
+}
+```
+
+`getActions` takes an optional `fromActionHash` parameter which is the `actionsHash` of the last action that was processed. If `fromActionHash` is not provided, `getActions` will fetch all actions for the zkApp. `getActions` returns a list of actions that were emitted after the `fromActionHash` was processed.
```
Comment (jasongitmail) [inline docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx:136]: `getActions` > `getActions()`

> `getActions()` accepts an object an optional `fromActionHash` property, where the value is the `actionsHash` of the most recent action that was processed.

> "was processed" 
Could we clarify to be explicit about what this means? I.e. 'received by the zkApp account' or something like that?

File: docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx [hunk 6]
```diff
@@ -99,6 +99,42 @@ this.reducer.getActions({ fromActionsHash?: Field, endActionsHash?: Field }): A[
...
+
 ### Actions for concurrent state updates
 
 We imagine that one of the most important use cases for actions is to enable concurrent state updates. This is also why actions where originally added to the protocol.
```
Comment (jasongitmail) [inline docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx:140]: opportunity to reword CC @barriebyron 
I'd suggest stating the value prop, without any "we imagine"

File: docs/zkapps/advanced-snarkyjs/events.mdx [hunk 7]
```diff
@@ -61,6 +61,34 @@ class SmartContract {
 }
 ```
 
+### Fetching Events from an Archive Node
+
+Since emitted event data is only stored in transaction data, events will only be persisted by an Archive Node. To fetch events from an Archive Node, you can use the `archive` property on the `Mina.Network` function. For example:
```
Comment (jasongitmail) [inline docs/zkapps/advanced-snarkyjs/events.mdx:66]: Would suggest rewording the first sentence:

> Events are contained within zkApp transactions and persisted by Mina archive nodes.

Would reword the 2nd sentence similar to my suggestions in the earlier section.

File: docs/zkapps/advanced-snarkyjs/events.mdx [hunk 8]
```diff
@@ -61,6 +61,34 @@ class SmartContract {
...
+
+```ts
+let Network = Mina.Network({
```
Comment (jasongitmail) [inline docs/zkapps/advanced-snarkyjs/events.mdx:69]: const

File: docs/zkapps/advanced-snarkyjs/events.mdx [hunk 9]
```diff
@@ -61,6 +61,34 @@ class SmartContract {
...
+  mina: '', // Use https://proxy.berkeley.minaexplorer.com/graphql or https://api.minascan.io/node/berkeley/v1/graphql
+  archive: '', // Use https://api.minascan.io/archive/berkeley/v1/graphql/ or https://archive.berkeley.minaexplorer.com/
+});
+Mina.setActiveInstance(Network);
+```
+
+By setting the `archive` property to one of the listed endpoints, SnarkyJS will use one of the selected endpoints to fetch events. For example, let us assume that a zkApp has already emitted events and has broadcasted the transaction. To fetch events for a zkApp, you can use the `fetchEvents` function:
```
Comment (jasongitmail) [inline docs/zkapps/advanced-snarkyjs/events.mdx:76]: Would rephrase b/c the URL isn't the actor. I.e. it's fetched from an endpiont, the endpoint doesn't do it.

`fetchEvents()`

File: docs/zkapps/advanced-snarkyjs/events.mdx [hunk 10]
```diff
@@ -61,6 +61,34 @@ class SmartContract {
...
+
+```ts
+let zkapp = new SimpleZkapp(address);
+// fetches all events from zkapp starting block 0
+let events = await zkapp.fetchEvents(UInt32.from(0));
```
Comment (jasongitmail) [inline docs/zkapps/advanced-snarkyjs/events.mdx:81]: Could comment out one of the lines and start both with `let`, since it's an alternate example and both lines won't be used simultaneously.

fetches > fetch

File: docs/zkapps/advanced-snarkyjs/events.mdx [hunk 11]
```diff
@@ -61,6 +61,34 @@ class SmartContract {
...
+// fetches all events starting at block 560 and ending at block 600
+events = await zkapp.fetchEvents(UInt32.from(560), UInt32.from(600));
+// fetches all events for a given address
+const fetchedEvents = await fetchEvents({
+  publicKey: 'B62qrfn5xxChtPGJne9HuDJZ4ziWVgWxeL3hntGBqMmf45p4hudo3tw',
+});
+```
+
+`getEvents` takes in a starting block and an ending block. If the ending block is not provided, it will fetch all events up to the latest block. If the starting block is not provided, it will fetch all events from the beginning of the zkApp's history.
```
Comment (jasongitmail) [inline docs/zkapps/advanced-snarkyjs/events.mdx:90]: should be `fetchEvents()`, not `getEvents()` right?

Comment (barriebyron) [inline docs/zkapps/advanced-snarkyjs/events.mdx:90]:
```suggestion
`getEvents` takes in a starting block and an ending block. 

- If the ending block is not provided, it fetches all events up to the latest block. 
- If the starting block is not provided, it fetches all events from the beginning of the zkApp's history.
```

File: docs/zkapps/advanced-snarkyjs/events.mdx (hunk 9 above)
Comment (barriebyron) [inline docs/zkapps/advanced-snarkyjs/events.mdx:76]:
```suggestion
By setting the `archive` property to one of the listed endpoints, SnarkyJS uses one of the selected endpoints to fetch events. For example, if a zkApp has already emitted events and has broadcasted the transaction. To fetch events for a zkApp, you can use the `fetchEvents` function:
```

**Review by barriebyron — CHANGES_REQUESTED** [review]
@MartinMinkov  I had a good time learning about actions and events... you can teach me a lot about our tech with this PR

I suggested editorial updates... have you used the GitHub suggesting changes feature before? You can commit suggestions one at a time or in batch. 

Let's talk through any changes that make you wonder, we can teach each other!

File: docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx (hunk 1 above)
Comment (jasongitmail) [reply docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx:104]: So I interpret this as saying it's an optional property on the object used by `Mina.network()`, but the two methods you mentioned will fail at either build or runtime in that scenario. Is that correct? If so, can you create an issue on the Custom ESLint Plugin to identify if this scenario exists, so we can warn about it? Goal is to give devs feedback earlier.

File: docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx [hunk 12]
```diff
@@ -99,9 +99,63 @@ this.reducer.getActions({ fromActionsHash?: Field, endActionsHash?: Field }): A[
 
 Right now, `getActions` is available for [testing with `LocalBlockchain`](/zkapps/how-to-test-a-zkapp).
 
+### Fetching Actions from an Archive Node
+
+Since Mina nodes do not store historical network information and actions are not kept on-chain, emitted actions are only preserved in the Archive Node. Consequently, a zkApp must fetch its previously emitted actions on the Mina network from an Archive Node.
+
+As the Archive Node is essentially a database, it does not offer a user-friendly API that SnarkyJS can readily utilize to fetch actions. To address this issue, a GraphQL server was developed to retrieve network information such as actions and actions from the Archive Node. SnarkyJS will use this GraphQL server to fetch actions from the Archive Node.
```
Comment (barriebyron) [inline docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx:106]:
```suggestion
As the Archive Node is essentially a database, it does not offer a user-friendly API that SnarkyJS can readily utilize to fetch actions. To address this issue, a GraphQL server was developed to retrieve network information such as actions and actions from the Archive Node. SnarkyJS uses this GraphQL server to fetch actions from the Archive Node.
```

Comment (barriebyron) [reply docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx:106]: even when future events are mentioned, style guidance is to use present tense https://developers.google.com/style/tense please apply throughout

File: docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx [hunk 13]
```diff
@@ -99,9 +99,63 @@ this.reducer.getActions({ fromActionsHash?: Field, endActionsHash?: Field }): A[
...
+
+To configure SnarkyJS to use an Archive Node to fetch actions, use the `archive` property in the `Mina.Network` function. The `archive` property must be specified in the `Mina.Network` function to fetch actions from an Archive Node; otherwise, an error will occur.
```
Comment (barriebyron) [inline docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx:108]:
```suggestion
To configure SnarkyJS to use an Archive Node to fetch actions, use the `archive` property in the `Mina.Network` function. The `archive` property must be specified in the `Mina.Network` function to fetch actions from an Archive Node; otherwise, an error occurs.
```

File: docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx [hunk 14]
```diff
@@ -99,9 +99,63 @@ this.reducer.getActions({ fromActionsHash?: Field, endActionsHash?: Field }): A[
...
+      },
+      { state: counter, actionsHash }
+    );
+  }
+}
+```
+
+The `getActions()` function accepts an optional `fromActionHash` parameter. This `actionsHash` is derived from processing Account Updates and indicates the starting point for your zkApp to begin processing actions. If `fromActionHash` is not provided, `getActions()` get all actions for the zkApp, since the first one. The function returns a list of actions emitted after the specified `fromActionHash` has been processed.
+
+To see a end-to-end example of zkApp fetching actions from a running network and including them into your zkApp, [see this example](https://github.com/o1-labs/snarkyjs/blob/main/src/examples/zkapps/voting/voting.ts).
```
Comment (barriebyron) [inline docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx:154]:
```suggestion
To see an end-to-end example of zkApp fetching actions from a running network and including them into your zkApp,  see this [voting app example](https://github.com/o1-labs/snarkyjs/blob/main/src/examples/zkapps/voting/voting.ts).
```

Comment (barriebyron) [reply docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx:154]: we have an initiative to use descriptive link text, per https://developers.google.com/style/link-text

File: docs/zkapps/advanced-snarkyjs/events.mdx [hunk 15]
```diff
@@ -61,6 +61,63 @@ class SmartContract {
 }
 ```
 
+### Fetching Events from an Archive Node
+
+Since Mina nodes do not store historical network information and events are not kept on-chain, emitted events are only preserved in the Archive Node. Consequently, a zkApp must fetch its previously emitted events on the Mina network from an Archive Node.
```
Comment (barriebyron) [inline docs/zkapps/advanced-snarkyjs/events.mdx:66]:
```suggestion
Since Mina nodes do not store historical network information and events are not kept on-chain, emitted events are preserved only in the Archive Node. Consequently, a zkApp must fetch its previously emitted events on the Mina network from an Archive Node.
```

File: docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx (hunk 12 above)
Comment (barriebyron) [reply docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx:106]: hit up the present tense throughout, thanks!

Comment (barriebyron) [inline docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx:106]: I see this content is repeated in another topic, let's sync up and talk about how we can repurpose the archive node description in one place and then have that place come to these new (awesome, btw!) topics

**Review by barriebyron — COMMENTED** [review]
@MartinMinkov  well done!! This content is clear, concise, descriptive, and written for a developer audience that is ready to implement this feature

I made a few insubstantial editorial suggestions.

On the larger side, I'd like to talk more about the archive node architecture and how we can put that information in a single place that makes more sense... maybe near https://docs.minaprotocol.com/node-operators/archive-node#using-the-archive-node or similar? (I have a lot to learn, teach me as we go)

File: docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx (hunk 12 above)
Comment (nicc) [reply docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx:106]: Note "actions and actions". Might be intended as "events as actions"?

Comment (barriebyron) [reply docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx:106]: Good catch

Comment (jasongitmail) [reply docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx:106]: I'm going to suggest a bigger change here 1.) to avoid confusion over database vs API, and 2.) to highlight the ability to store data using any datastore, given past experience where community wrongly read into old docs on this specific aspect, and we rewrite the docs to be more clear that any data store can be used.

> The Mina archive node offers the ability for operators to store data using any data store of their choice, however, it does not currently expose a built-in API to query this data publicly. To address this, the [Mina Archive Node API](https://github.com/o1-labs/archive-node-api) was developed and can be run by archive node operators to provide a GraphQL API from which zkApp smart contracts can retrieve events and actions.

File: docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx (hunk 13 above)
Comment (jasongitmail) [reply docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx:108]: Building on Barrie's:

> If your smart contract needs to fetch actions from an archive node, simply provide an `archive` property within the object passed into `Mina.Network({mina: '...', archive: '...'})`. Its value should be the URL for the GraphQL API service that you want to use. If this property does not exist, an error will occur.

Future tense seems appropriate in this case (`an error will occur`) given it's an "if/then", no?

File: docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx [hunk 16]
```diff
@@ -99,9 +99,63 @@ this.reducer.getActions({ fromActionsHash?: Field, endActionsHash?: Field }): A[
...
+
+```sh
+https://api.minascan.io/archive/berkeley/v1/graphql
+```
+
+```sh
+https://archive.berkeley.minaexplorer.com
+```
+
+For example, to use the `archive` property, see the following example code:
```
Comment (jasongitmail) [inline docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx:120]:
```suggestion
For example:
```

File: docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx [hunk 17]
```diff
@@ -99,9 +99,63 @@ this.reducer.getActions({ fromActionsHash?: Field, endActionsHash?: Field }): A[
...
+
+```ts
+const Network = Mina.Network({
+  mina: '', // Use https://proxy.berkeley.minaexplorer.com/graphql or https://api.minascan.io/node/berkeley/v1/graphql
+  archive: '', // Use https://api.minascan.io/archive/berkeley/v1/graphql/ or https://archive.berkeley.minaexplorer.com/
+});
+Mina.setActiveInstance(Network);
+```
+
+By setting the `archive` property to one of the listed endpoints, SnarkyJS uses the selected endpoint to fetch events. For instance, if a zkApp has already emitted events and broadcasted the transaction, you can use the `getActions()` function to retrieve actions for a zkApp:
```
Comment (jasongitmail) [inline docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx:130]: Could we simply this to? 

> Within your smart contract, you can now use `getActions()` to retrieve actions emitted by your smart contract as part of previous transactions.

File: docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx [hunk 18]
```diff
@@ -99,9 +99,63 @@ this.reducer.getActions({ fromActionsHash?: Field, endActionsHash?: Field }): A[
...
+      (state: Field, action: Field) => {
+        ...
+      },
+      { state: counter, actionsHash }
+    );
+  }
+}
+```
+
+The `getActions()` function accepts an optional `fromActionHash` parameter. This `actionsHash` is derived from processing Account Updates and indicates the starting point for your zkApp to begin processing actions. If `fromActionHash` is not provided, `getActions()` get all actions for the zkApp, since the first one. The function returns a list of actions emitted after the specified `fromActionHash` has been processed.
```
Comment (jasongitmail) [inline docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx:152]:
```suggestion
By default, `getActions()` retrieves all actions from the very first ever emitted for that zkApp account. However, you can provide an object containing an optional property named `fromActionHash` where the value is an `actionsHash` to indicate the starting point of the actions to be retrieved and processed by your smart contract. Note that actions _after_ the specified hash will be retrieved—i.e. it is not inclusive.
```

@MartinMinkov From your text, it appears that the specified actionHash is excluded from the returned results. Is that correct? That's somewhat surprising behavior, no? Is there is a reason it can't or shouldn't be inclusive?

File: docs/zkapps/advanced-snarkyjs/events.mdx [hunk 19]
```diff
@@ -61,6 +61,63 @@ class SmartContract {
 }
 ```
 
+### Fetching Events from an Archive Node
```
Comment (jasongitmail) [inline docs/zkapps/advanced-snarkyjs/events.mdx:64]: My suggestions above can be looked at and adapted down here too, if accepted, given this section mirrors the above.

Comment (barriebyron) [reply docs/zkapps/advanced-snarkyjs/events.mdx:64]: @jasongitmail  see the updated PR description about having a single source of the archive node content... as best practice, we do not want to mirror or duplicate content

File: docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx (hunk 18 above)
Comment (barriebyron) [reply docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx:152]: It seems awkward at first, but I still promote present tense, even for future actions
In this case
will be retrieved >> is retrieved

https://developers.google.com/style/tense

File: docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx [hunk 20]
```diff
@@ -99,9 +99,61 @@ this.reducer.getActions({ fromActionsHash?: Field, endActionsHash?: Field }): A[
...
+      (state: Field, action: Field) => {
+        ...
+      },
+      { state: counter, actionsHash }
+    );
+  }
+}
+```
+
+By default, `getActions()` retrieves all actions from the very first ever emitted for that zkApp account. However, you can provide an object containing an optional property named `fromActionHash` where the value is an `actionsHash` to indicate the starting point of the actions to be retrieved and processed by your smart contract. Note that actions _after_ the specified hash will be retrieved—i.e. it is not inclusive.
```
[the remark this answers is not in the document]
Comment (barriebyron) [reply docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx:150]: I wasn't clear about suggesting that we change "containing" to "that contains" 
Is there a technical difference? If not, let's use "that contains"

File: docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx [hunk 21]
```diff
@@ -99,9 +99,61 @@ this.reducer.getActions({ fromActionsHash?: Field, endActionsHash?: Field }): A[
...
+
+To see a end-to-end example of zkApp fetching actions from a running network and including them into your zkApp, see this voting app example in the [voting.ts](https://github.com/o1-labs/snarkyjs/blob/main/src/examples/zkapps/voting/voting.ts) file in examples provided in the codebase repo.
```
Comment (barriebyron) [inline docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx:152]:
```suggestion
For an end-to-end example of zkApp fetching actions from a running network and including them into your zkApp, see this voting app example in the [voting.ts](https://github.com/o1-labs/snarkyjs/blob/main/src/examples/zkapps/voting/voting.ts) file in examples provided in the codebase repo.
```

File: docs/zkapps/advanced-snarkyjs/events.mdx [hunk 22]
```diff
@@ -61,6 +61,58 @@ class SmartContract {
...
+
+Since Mina nodes do not store historical network information and events are not kept on-chain, emitted events are preserved only in the archive node. To see how to work with events within your zkApp, refer to our [Using the Archive Node with SnarkyJS](../../node-operators/archive-node#using-the-archive-node-with-snarkyjs) section.
+
+If your smart contract needs to fetch events from an archive node, simply provide an `archive` property within the object passed into `Mina.Network({mina: '...', archive: '...'})`. Its value should be the URL for the GraphQL API service that you want to use. If this property does not exist, an error will occur.
```
Comment (barriebyron) [inline docs/zkapps/advanced-snarkyjs/events.mdx:68]: same feedback as above... we can work together to find the best way forward to omit duplicate architectural content

**Review by barriebyron — COMMENTED** [review]
@MartinMinkov  you are raising the bar on the Archive Node content! Well done.

Let's work together to find a clear way forward with the content, so that we find the best way to present it our readers while minimizing churn for you

File: docs/zkapps/advanced-snarkyjs/events.mdx [hunk 23]
```diff
@@ -61,6 +61,58 @@ class SmartContract {
...
+
+To see a end-to-end example of zkApp fetching events from a running network, see this voting app example in the [run_berkeley.ts](https://github.com/o1-labs/snarkyjs/blob/main/src/examples/zkapps/voting/run_berkeley.ts) file in examples provided in the codebase repo.
+
 ### Events vs Actions
 
-Events and [Actions](actions-and-reducer) are two distinct mechanisms for logging information alongside a transaction. Events are not meant for use within proofs directly as they can't be predicated on inside proofs. Events are used to signal to UIs, but can also be used for reconstructing Merkle trees as we see above. Actions, on the other hand, can be accessed within provable code within a smart contract using a Reducer, because a commitment to the relevant actions (i.e. the "action state") is stored on the zkApp account. Neither Events nor Actions are stored directly in the Mina ledger and exist only in the transaction and on archive nodes.
+Events and [Actions](actions-and-reducer) are two distinct mechanisms for logging information alongside a transaction.
+
```
[the remark this answers is not in the document]
Comment (jasongitmail) [reply docs/zkapps/advanced-snarkyjs/events.mdx:113]: These could be added to the glossary. Actions & events are features of Mina Protocol.

File: docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx (hunk 17 above)
Comment (jasongitmail) [reply docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx]: @barriebyron Goal here is that they have to have done the preceeding step in order for this to be possible. That's the goal of now. Maybe you can suggest an alternative?

**Comment by nicc** [comment]
Context buddy: @barriebyron  / @nicc

File: docs/node-operators/archive-node.mdx [hunk 24]
```diff
@@ -9,12 +9,17 @@ A new version of Mina Docs is coming soon! This page will be rewritten.
...
 
-
 # Archive Node
 
 Mina nodes are succinct by default, meaning they don't need to maintain historical information about the network, block, or transactions.
 For some usecases, it is useful to maintain this historical data, which you can do by running an **Archive Node.**
 
+:::tip
+
+A zkApp can retrieve events and actions from one or more Mina archive nodes. If your smart contract needs to fetch events and actions from an archive node, see [How to Fetch Events and Actions](../zkapps/advanced-snarkyjs/#fetch-events-and-actions).
```
Comment (jasongitmail) [inline docs/node-operators/archive-node.mdx:19]: Just fyi. We can currently only fetch from one archive node, but we have plans to fetch from more than one simultaneously for redundancy, in the next sprint.

This is probably ok as is.

File: docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx [hunk 25]
```diff
@@ -101,12 +107,8 @@ Right now, `getActions` is available for [testing with `LocalBlockchain`](/zkapp
...
 
-We also have a [full code example](https://github.com/o1-labs/snarkyjs/blob/main/src/examples/zkapps/reducer/reducer.ts) which demonstrates this pattern. Leveraging `Reducer.reduce`, it takes only about 30 lines of code to build a zkApp which handles concurrent state updates.
-
-### Events vs Actions
-
-[Events](events) and Actions are two distinct mechanisms for logging information alongside a transaction. Events are not meant for use within proofs directly as they can't be predicated on inside proofs. Events are used to signal to UIs, but can also be used for reconstructing merkle trees. Actions, on the other hand, can be accessed within provable code via Reducers as we see above. Both Events and Actions are not stored in the ledger and only exist on the transaction.
+We also have a [full code example](https://github.com/o1-labs/snarkyjs/blob/main/src/examples/zkapps/reducer/reducer.ts) which demonstrates this pattern. Leveraging `Reducer.reduce`, it takes only about 30 lines of code to build a zkApp which handles concurrent state updates.
```
Comment (jasongitmail) [inline docs/zkapps/advanced-snarkyjs/actions-and-reducer.mdx]: `Reducer.reduce()`

File: docs/zkapps/advanced-snarkyjs/events.mdx [hunk 26]
```diff
@@ -49,7 +49,13 @@ Some other important facts about events:
...
 - You can't refer to previously emitted events in a smart contract, because there is no way of proving that the events you refer to are actually the events emitted by that contract.
 
-This is all you need to know about events! Think of them as a convenience feature -- a light-weight way of attaching information about your smart contract execution, which would otherwise get lost. Don't treat them as fully-fledged storage which can be safely accessed in smart contracts.
+:::tip
+
+A zkApp can retrieve events and actions from one or more Mina archive nodes. If your smart contract needs to fetch events and actions from an archive node, see [How to Fetch Events and Actions](fetch-events-and-actions).
+
+:::
+
+This is all you need to know about events! Think of them as a convenience feature -- a light-weight way of attaching information about your smart contract execution, which would otherwise get lost. Don't treat them as fully-fledged storage that can be safely accessed in smart contracts.
```
Comment (jasongitmail) [inline docs/zkapps/advanced-snarkyjs/events.mdx:58]: should we use an actual em dash here? `—`

Comment (jasongitmail) [reply docs/zkapps/advanced-snarkyjs/events.mdx:58]: Suggestion. 'Getting lost' seems negative and I wonder if this suggestion gets to the intent better:

`, which would otherwise get lost`   >  `, to expose the event to the outside world such as your UI`

**Review by jasongitmail — COMMENTED** [review]
Left some literary comments, but it might be worth noting that this will change how we have approached the Advanced SnarkyJS section away from each page being focused on a specific feature. So the tradeoff is organizational approach vs written text redundancy.

**Review by barriebyron — APPROVED** [review]

**Comment by barriebyron** [comment]
@jasongitmail  can you teach me why this PR is failing?

**Comment by MartinMinkov** [comment]
> @jasongitmail can you teach me why this PR is failing?

There is a build error, caused by broken links! :grimacing: 
This is the error log:
```bash
On source page path = /node-operators/archive-node:
--
07:48:12.797 | -> linking to ../zkapps/advanced-snarkyjs#fetch-events-and-actions (resolved as: /zkapps/advanced-snarkyjs)
```

**Comment by barriebyron** [comment]
Preview link of the new file https://docs2-git-feat-add-docs-for-events-and-actions-minadocs.vercel.app/zkapps/advanced-snarkyjs/fetch-events-and-actions

**Comment by barriebyron** [comment]
> Left some literary comments, but it might be worth noting that this will change how we have approached the Advanced SnarkyJS section away from each page being focused on a specific feature. So the tradeoff is organizational approach vs written text redundancy. Something for us to consider.

@jasongitmail @nicc 
We want to avoid duplicating content for multiple reasons: 
- duplicate content confuses search engines
- putting content in more than one place adds complexity to updates and maintenance
- organizing content for positive UX and successful DevX is our goal, so let's find a way to address how to best serve our readers

 Let's sync on this content and examine in the broader content architecture improvements for all SnarkyJS docs proposed in #357 https://github.com/o1-labs/docs2/issues/357

**Review by jasongitmail — APPROVED** [review]
Approving per team chat so this content is live. We have a discussion planned around docs structure later this week and can reorganize afterward as needed.
